grass/kernel.c

```diff
@@ -174,12 +181,32 @@ static void proc_syscall() {
 
     switch (type) {
     case SYS_RECV:
-        proc_recv(sc);
+        rc = proc_recv(sc);
         break;
     case SYS_SEND:
-        proc_send(sc);
+        rc = proc_send(sc);
+        break;
+    case SYS_WAIT:
+        waiting++;
+        if (wakeup) rc = waiting = wakeup = 0;
+        break;
+    case SYS_EXIT:
+        proc_free(curr_pid);
+        if (waiting) wakeup++;
         break;
     default:
         FATAL("proc_syscall: got unknown syscall type=%d", type);
     }
+
+    if (rc == -1)
+        sc->type = type; // Failure, Retry Request
```

Comment:
I don't understand why we need to set `sc->type = type`. Did we modify `sc->type` anywhere above?

Reply:
From my understanding, we initially set `sc->type` to `SYS_UNUSED` in `proc_syscall` so when returning to user space, we exit the while loop `while (sc->type != SYS_UNUSED)`. On a failed system call, we should keep the type of the system call the same.

To make the system call type change clearer, I moved the update of `sc->type` to only be within the else branch of `if (rc == -1)`